curvefs/src/client/block_device_client.h

```diff
@@ -25,80 +25,148 @@
-     *
-     * @return error code
+     * @param[in] options the options for client
+     * @return error code (CURVEFS_ERROR:*)
      */
-    virtual CURVEFS_ERROR Init(const BlockDeviceClientOptions &options) = 0;
+    virtual CURVEFS_ERROR Init(const BlockDeviceClientOptions& options) = 0;
 
     /**
      * @brief Uninitailize client
-     *
-     * @return error code
      */
-    virtual CURVEFS_ERROR UnInit() = 0;
+    virtual void UnInit() = 0;
 
     /**
-     * @brief read
-     *
-     * @param buf read buffer
-     * @param offset read start offset
-     * @param length read length
-     *
-     * @return error code
+     * @brief Open file specify by filename and owner
+     * @param[in] filename
+     * @param[in] owner owner for filename
+     * @return error code (CURVEFS_ERROR:*)
      */
-    virtual CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) = 0;
+    virtual CURVEFS_ERROR Open(const std::string& filename,
+                               const std::string& owner) = 0;
 
+    /**
+     * @brief Close the fd which init by Open()
+     * @return error code (CURVEFS_ERROR:*)
+     */
+    virtual CURVEFS_ERROR Close() = 0;
+
+    /**
+     * @brief Get file status
+     * @param[in] filename
+     * @param[in] owner owner for filename
+     * @param[out] statInfo the struct for file status
+     * @return error code (CURVEFS_ERROR:*)
+     */
+    virtual CURVEFS_ERROR Stat(const std::string& filename,
+                               const std::string& owner,
+                               BlockDeviceStat* statInfo) = 0;
+
+    /**
+     * @brief Read from fd which init by Open()
+     * @param[in] buf read buffer
+     * @param[in] offset read start offset
+     * @param[in] length read length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
+     */
+    virtual CURVEFS_ERROR Read(char* buf, off_t offset, size_t length) = 0;
 
     /**
-     * @brief write
-     *
-     * @param fd file descriptor
-     * @param buf write buffer
-     * @param offset write start offset
-     * @param length write length
-     *
-     * @return error code
+     * @brief Write to fd which init by Open()
+     * @param[in] buf write buffer
+     * @param[in] offset write start offset
+     * @param[in] length write length
+     * @return error code (CURVEFS_ERROR:*)
+     * @note the offset and length maybe not aligned
      */
     virtual CURVEFS_ERROR Write(
         const char* buf, off_t offset, size_t length) = 0;
 };
 
 class BlockDeviceClientImpl : public BlockDeviceClient {
  public:
-    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient) :
-        fileClient_(fileClient) {}
-    virtual ~BlockDeviceClientImpl() {}
+    explicit BlockDeviceClientImpl(std::shared_ptr<FileClient> fileClient)
+        : fileClient_(fileClient) {}
 
-    CURVEFS_ERROR Init(const BlockDeviceClientOptions &options) override;
+    BlockDeviceClientImpl() = default;
```

Comment:
why add an extra default constructor?

Reply:
> why add an extra default constructor?

This is an unnecessary constructor, i will delete it.

Reply:
I need a constructor as follows, please see my pr: https://github.com/opencurve/curve/pull/402/files#diff-97023361921fe6f18ece5ed622166ee6036bab57b9553e4cdebfa094dfe7e898

And the "fd_" & "isOpen_" shoule be initailize in initailize list !

BlockDeviceClientImpl()  
      : fileClient_(std::make_shared<FileClient>()),
        fd_(-1),
        isOpen_(false)  {}

Reply:
> I need a constructor as follows, please see my pr: https://github.com/opencurve/curve/pull/402/files#diff-97023361921fe6f18ece5ed622166ee6036bab57b9553e4cdebfa094dfe7e898
> 
> And the "fd_" & "isOpen_" shoule be initailize in initailize list !
> 
> BlockDeviceClientImpl()
> : fileClient_(std::make_shared()),
> fd_(-1),
> isOpen_(false) {}

done